webapi/CopilotChat/Skills/ChatSkills/ExternalInformationSkill.cs

```diff
@@ -118,7 +118,25 @@ public class ExternalInformationSkill
         {
             // Create a plan and set it in context for approval.
             var contextString = string.Join("\n", context.Variables.Where(v => v.Key != "userIntent").Select(v => $"{v.Key}: {v.Value}"));
-            Plan plan = await this._planner.CreatePlanAsync($"Given the following context, accomplish the user intent.\nContext:\n{contextString}\nUser Intent:{userIntent}");
+            Plan? plan = null;
+            int maxRetries = 1;
+
+            do
+            { // TODO: [Issue #2256] Remove retry logic once Core team stabilizes planner
+                try
+                {
+                    plan = await this._planner.CreatePlanAsync($"Given the following context, accomplish the user intent.\nContext:\n{contextString}\nUser Intent:{userIntent}");
+                }
+                catch (PlanningException e) when (e.ErrorCode == PlanningException.ErrorCodes.InvalidPlan && this._planner.PlannerOptions!.AllowRetriesOnInvalidPlans)
+                {
+                    if (maxRetries-- > 0)
+                    {
+                        // Retry plan creation if LLM returned response that doesn't contain valid plan (invalid XML or JSON).
+                        context.Log.LogWarning("Retrying CreatePlan on error: {0}", e.Message);
+                        continue;
+                    }
```

Comment:
it seems like we need a `break` after this conditional to prevent an infinite loop?

Reply:
I don't think we should ever run into that case because any exception or error in the block will force the CreatePlanAsync method to bail out, but I added a break as an extra safeguard

Reply:
Actually, I changed it to throw to be more accurate, and formatter confirmed that we should ever run into that case because any exception will throw and propagate up so removing.


<img width="691" alt="image" src="https://github.com/microsoft/chat-copilot/assets/125500434/6303291a-3027-4972-a6f7-b0344503866c">

Reply:
gotcha, so the Planner is guaranteed to return a valid plan at some point?

Reply:
It's either guaranteed to return a plan object (not necessarily valid) or throw an error